File: 008. Node Depths/NodeDepthsRecursive.java
import java.util.*;

class NodeDepthsRecursive {

    public static int nodeDepths(BinaryTree root) {
        return nodeDepths(root, 0);
    }

    public static int nodeDepths(BinaryTree node, int depth) {
        if (node == null)
            return 0;
        return depth + nodeDepths(node.left, depth + 1) + nodeDepths(node.right, depth + 1);
    }

    static class BinaryTree {
        int value;
        BinaryTree left;
        BinaryTree right;

        public BinaryTree(int value) {
            this.value = value;
            left = null;
            right = null;
        }
    }
}

File: 8. Node Depths/NodeDepthsRecursive.kt
package com.algoexpert.program

open class BinaryTree(value: Int) {
    var value = value
    var left: BinaryTree? = null
    var right: BinaryTree? = null
}

fun nodeDepths(root: BinaryTree): Int {
    return nodeDepths(root, 0)
}

fun nodeDepths(node: BinaryTree?, depths: Int): Int {
    if (node == null) {
        return 0
    }
    return depths + nodeDepths(node.left, depths + 1) + nodeDepths(node.right, depths + 1)
}
